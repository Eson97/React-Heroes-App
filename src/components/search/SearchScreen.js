import React, { useMemo } from 'react';
import queryString from 'query-string';

import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {


    const location = useLocation();
    const { q: query = '' } = queryString.parse(location.search);

    const [formValues, handleInputChange, reset] = useForm({ inputSearch: query });
    const { inputSearch } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(query), [query]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${inputSearch}`);
        reset();
    }

    return (
        <div>
            <form onSubmit={handleSearch}>
                <div className="input-group mb-3" style={{ maxWidth: 520 }}>
                    <input
                        type="text"
                        className="form-control"
                        autoComplete="off"
                        name="inputSearch"
                        value={inputSearch}
                        onChange={handleInputChange}
                        placeholder="Find your Hero!!"
                    />
                    <button className="btn btn-outline-secondary" type="submit">Search...</button>
                </div>
            </form>
            <hr />

            {
                ((query !== '') && heroesFiltered.length === 0) &&
                <div className="alert alert-danger">Hero not found!</div>
            }

            <div className="row">
                {
                    heroesFiltered.map(hero => (
                        <div className="col-6 col-md-4 animate__animated animate__fadeIn" key={hero.id}>
                            <HeroCard {...hero} />
                        </div>
                    ))
                }
            </div>

        </div>
    )
}
