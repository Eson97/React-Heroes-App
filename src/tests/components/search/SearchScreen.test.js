import { act } from "@testing-library/react"
import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { SearchScreen } from "../../../components/search/SearchScreen"



describe('Pruebas en SearchScreen', () => {

    test('should show correctly with default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}> {/*asigna la ruta inicial*/}
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').exists()).toBeFalsy();
        expect(wrapper.find('.btn-outline-secondary').exists()).toBeTruthy();
    })

    test('should display a BatmanÂ´s HeroCard', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}> {/*asigna la ruta inicial*/}
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('HeroCard').exists()).toBeTruthy()
    })

    test('should display an error box', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}> {/*asigna la ruta inicial*/}
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('.alert-danger').exists()).toBeTruthy();
        expect(wrapper.find('.alert').text().trim()).toBe('Hero not found!');
    })

    test('should call PUSH from history', () => {

        const historyMock = {
            push: jest.fn()
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}> {/*asigna la ruta inicial*/}
                <Route path='/search' component={() => <SearchScreen history={historyMock} />} />
            </MemoryRouter>
        )
        wrapper.find('input').simulate('change', {
            target: {
                name: 'inputSearch',
                value: 'batman'
            }
        })

        act(() => {
            wrapper.find('form').prop('onSubmit')({
                preventDefault() { }
            })
        })

        expect(historyMock.push).toHaveBeenCalledWith('?q=batman')
    })

})
