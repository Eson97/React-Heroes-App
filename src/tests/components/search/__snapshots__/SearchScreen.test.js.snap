// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen should show correctly with default values 1`] = `
Array [
  " ",
  <div>
    <form
      onSubmit={[Function]}
    >
      <div
        className="input-group mb-3"
        style={
          Object {
            "maxWidth": 520,
          }
        }
      >
        <input
          autoComplete="off"
          className="form-control"
          name="inputSearch"
          onChange={[Function]}
          placeholder="Find your Hero!!"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-secondary"
          type="submit"
        >
          Search...
        </button>
      </div>
    </form>
    <hr />
    <div
      className="row"
    />
  </div>,
]
`;
